File: 小可爱demo/heap.ts
export class HeapNode<T> {
    constructor(public weight: number, public payload: T) {
        this.weight = weight;
        this.payload = payload;
    }
}
export function swap<T>(arr: T[], i: number, j: number): void {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

abstract class HeapFace<T> {
    protected root: [null,...HeapNode<T>[]] = [null];
    protected size = 0;
    abstract offer(node:HeapNode<T>):void;
    abstract poll():HeapNode<T>|null;
    abstract peek():HeapNode<T>|null;
    abstract heapifyUp():void;
    abstract heapifyDown():void;
}

class Heap<T> extends HeapFace<T> {
    constructor(){
        super();
    }
    offer(node: HeapNode<T>): void {
        this.root.push(node);
        this.heapifyUp();
        this.size++;
    }
    poll(): HeapNode<T>|null {
        let result = null;
        if(this.root.length > 1) {
           swap(this.root, 1, this.root.length - 1);
           result = this.root.pop()!;
           this.heapifyDown();
           this.size--;
        }
        return result;
    }
    peek(): HeapNode<T>|null { return this.root[1] || null;}
    heapifyUp(): void {}
    heapifyDown(): void {}
}
//min head heap.
export class MinHeap<T> extends Heap<T> {
    constructor(){super();}
    heapifyUp(): void {
        const end = this.root.length - 1;
        let current = end;
        while(current > 1) {
           const parent = Math.floor(current / 2);
           if(this.root[current]!.weight < this.root[parent]!.weight) swap(this.root, current, parent);
           else break;
           current = parent;
        }
    }
    heapifyDown(): void {
        const start = 1,end = this.root.length - 1;
        let current = start;
        while((current<<1) <= end) {
            const left = current << 1;
            const right = left + 1;
            const next = right <= end && this.root[right]!.weight < this.root[left]!.weight ? right : left;
            if(this.root[current]!.weight <= this.root[next]!.weight) break;
            swap(this.root, current, next);
            current = next;
        }
    }
}
//max head heap
export class MaxHeap<T> extends Heap<T> {
    constructor(){super();}
    heapifyUp(): void {
        const end = this.root.length - 1;
        let current = end;
        while(current > 1) {
            const parent = Math.floor(current / 2);
            if(this.root[current]!.weight > this.root[parent]!.weight) swap(this.root, current, parent);
            else break;
            current = parent;
        }
    }
    heapifyDown(): void {
        const start = 1,end = this.root.length - 1;
        let current = start;
        while((current<<1) <= end) {
            const left = current << 1;
            const right = left + 1;
            const next = right <= end && this.root[right]!.weight > this.root[left]!.weight ? right : left;
            if(this.root[current]!.weight >= this.root[next]!.weight) break;
            swap(this.root, current, next);
            current = next;
        }
    }
}

File: ruststartup/leetcode/src/tool/heap.rs
use std::marker::PhantomData;

#[derive(Copy, Clone, PartialEq,Debug)]
pub struct HeapNode<T> {
   pub value: T,
   pub index: i32,
}
pub trait Heapify<T> {
    fn heapify_up(&self, data: &mut Vec<Option<HeapNode<T>>>);
    fn heapify_down(&self, data: &mut Vec<Option<HeapNode<T>>>);
}
pub struct Heap<T> {
    data: Vec<Option<HeapNode<T>>>,
    child: Box<dyn Heapify<T>>
}
impl<T> Heap<T>{
    pub fn new(child: Box<dyn Heapify<T>>) -> Self {
        Heap { data: vec![None],child }
    }
    pub fn offer(&mut self,elem: HeapNode<T>) {
        // 添加元素到堆中
        self.data.push(Some(elem));
        self.child.heapify_up(&mut self.data);
    }
    pub fn poll(&mut self)->Option<HeapNode<T>> {
        // 删除堆顶元素
        if self.data.len() <= 1 {
            return None; // 堆为空
        }
        let len = self.data.len();
        self.data.swap(1,len-1);
        let removed = self.data.pop(); // 移除最后一个元素
        match removed {
            None => None,
            Some(node) => {
                self.child.heapify_down(&mut self.data); // 调整堆结构
                node
            }
        }
    }
}

pub struct MinHeap<T> {
    pub _marker: PhantomData<T>,
}
impl<T> Heapify<T> for MinHeap<T> {
    fn heapify_up(&self, data: &mut Vec<Option<HeapNode<T>>>) {//向上调整
        let len = data.len();
        let mut start = len - 1;
        while start > 1 {
            let parent = start / 2;
            if data[start].as_ref().unwrap().index < data[parent].as_ref().unwrap().index {
                data.swap(start, parent);
                start = parent;
            } else {
                break;
            }
        }
    }
    fn heapify_down(&self, data: &mut Vec<Option<HeapNode<T>>>) {//向下调整
        let mut start = 1;
        let len = data.len();
        while start*2 < len {
            let left = start * 2;
            let right = start * 2 + 1;
            let mut next = left;
            if right < len && data[right].as_ref().unwrap().index < data[left].as_ref().unwrap().index {
               next = right;
            }
            if data[start].as_ref().unwrap().index > data[next].as_ref().unwrap().index {
                data.swap(start, next);
                start = next;
            } else {
                break;
            }
       }
    }
}
pub struct MaxHeap<T> {
    pub _marker: PhantomData<T>,
}
impl<T> Heapify<T> for MaxHeap<T> {
    fn heapify_up(&self, data: &mut Vec<Option<HeapNode<T>>>) {//向上调整
        let len = data.len();
        let mut start = len - 1;
        while start > 1 {
            let parent = start / 2;
            if data[start].as_ref().unwrap().index > data[parent].as_ref().unwrap().index {
                data.swap(start, parent);
                start = parent;
            } else {
                break;
            }
        }
    }
    fn heapify_down(&self, data: &mut Vec<Option<HeapNode<T>>>) {//向下调整
        let mut start = 1;
        let len = data.len();
        while start*2 < len {
            let left = start * 2;
            let right = start * 2 + 1;
            let mut next = left;
            if right < len && data[right].as_ref().unwrap().index > data[left].as_ref().unwrap().index {
               next = right;
            }
            if data[start].as_ref().unwrap().index < data[next].as_ref().unwrap().index {
                data.swap(start, next);
                start = next;
            } else {
                break;
            }
       }
    }

}